### PR TITLE
Allow configuring JPEG restart marker interval on save

### PR DESCRIPTION
**Pull request by bgilbert**
If a JPEG includes restart markers, the decoder doesn't have to give up if it finds data corruption, but can scan forward to the next restart marker and resume decoding from there.  Restart markers add overhead, but are useful in applications that want to support a limited form of error recovery or random access within a JPEG.

libjpeg allows specifying the marker interval either in MCU blocks or in MCU rows.  Support both, via separate parameters, rather than requiring callers to do the math.

### CODE REVIEW

File: src/libImaging/Jpeg.h
Comment (homm) [inline src/libImaging/Jpeg.h:88]: Why different types?

Comment (bgilbert) [reply src/libImaging/Jpeg.h:88]: Those are the types libjpeg uses.  Looking through the libjpeg source code, there doesn't seem to be a good reason for it, and I agree it seems pretty nonsensical.  I can switch over to unsigned if that's preferred.

**Comment by radarhere** [comment]
Just linking to some more information, https://github.com/libjpeg-turbo/libjpeg-turbo/blob/main/libjpeg.txt describes these features as
> unsigned int restart_interval
> int restart_in_rows
>    To emit restart markers in the JPEG file, set one of these nonzero. Set restart_interval to specify the exact interval in MCU blocks (samples in lossless mode).  Set restart_in_rows to specify the interval in MCU rows.  (If restart_in_rows is not 0, then restart_interval is set after the image width in MCUs is computed.) Defaults are zero (no restarts).  One restart marker per MCU row is often a good choice.  NOTE: the overhead of restart markers is higher in grayscale JPEG files than in color files, and MUCH higher in progressive JPEGs.  If you use restarts, you may want to use larger intervals in those cases.

File: src/libImaging/Jpeg.h
Comment (bgilbert) [reply src/libImaging/Jpeg.h]: Done.

**Comment by bgilbert** [comment]
Updated with feedback from https://github.com/bgilbert/Pillow/pull/2.

File: Tests/test_file_jpeg.py
Comment (radarhere) [inline Tests/test_file_jpeg.py:664]:
```suggestion
        assert len(re.findall(b"\xff[\xd0-\xd7]", out.getvalue())) == markers
```
This seems simpler.

**Review by radarhere — APPROVED** [review]

**Comment by radarhere** [comment]
Could you provide a short explanation for why you picked the names "restart_marker_blocks" and "restart_marker_rows" instead of continuing with libjpeg's names of "restart_interval" and "restart_in_rows"?

**Comment by bgilbert** [comment]
I find the libjpeg names inconsistent and confusing, and tried to pick better ones.  I'm open to other ideas though.

**Comment by bgilbert** [comment]
Sorry, I missed the merge commit when force-pushing.  The only changes to the PR are the `test_file_jpeg.py` ones in the [net diff](https://github.com/python-pillow/Pillow/compare/70d9cdf79e422118510e5c0344b570d1ae11f5f5..e5727224794620720f9b73098cce8c80b16ac304).

**Review by radarhere — APPROVED** [review]

**Comment by bgilbert** [comment]
Is this ready to go in?

**Comment by radarhere** [comment]
I've approved it, so I raise no objections if it were to be merged in.

I'm holding off on actually merging for two reasons
1. I'm not completely certain about renaming "restart_interval" and "restart_in_rows", as that is subjective, and maybe others have thoughts
2. In case @homm has any further thoughts in general, as he has offered some thoughts already

**Review by homm — APPROVED** [review]

**Pull request merged**
